graphs/bidirectional_a_star.py

```diff
@@ -0,0 +1,223 @@
+    >>> n = Node(1, 4, 3, 4, 2, None)
+    >>> n.calculate_heuristic()
+    2
+    >>> l = [k, n]
+    >>> n == l[0]
+    False
+    >>> l.sort()
+    >>> n == l[0]
+    True
+    """
+
+    def __init__(self, pos_x, pos_y, goal_x, goal_y, g_cost, parent):
+        self.pos_x = pos_x
+        self.pos_y = pos_y
+        self.pos = (pos_y, pos_x)
+        self.goal_x = goal_x
+        self.goal_y = goal_y
+        self.g_cost = g_cost
+        self.parent = parent
+        self.h_cost = self.calculate_heuristic()
+        self.f_cost = self.g_cost + self.h_cost
+
+    def calculate_heuristic(self) -> float:
+        """
+        The heuristic here is the Manhattan Distance
+        Could elaborate to offer more than one choice
+        """
+        dy = abs(self.pos_x - self.goal_x)
+        dx = abs(self.pos_y - self.goal_y)
+        return dx + dy
+
+    def __lt__(self, other):
+        return self.f_cost < other.f_cost
+
+
+class AStar:
+    def __init__(self, start, goal):
+        self.start = Node(start[1], start[0], goal[1], goal[0], 0, None)
+        self.target = Node(goal[1], goal[0], goal[1], goal[0], 99999, None)
+
+        self.open_nodes = [self.start]
+        self.closed_nodes = []
+
+        self.reached = False
+
+        self.path = [(self.start.pos_y, self.start.pos_x)]
+        self.costs = [0]
+
+    def search(self):
+        while self.open_nodes:
+            # Open Nodes are sorted using __lt__
+            self.open_nodes.sort()
+            current_node = self.open_nodes.pop(0)
+
+            if current_node.pos == self.target.pos:
+                self.reached = True
+                self.path = self.retrace_path(current_node)
+                break
+
+            self.closed_nodes.append(current_node)
+            successors = self.get_successors(current_node)
+
+            for child_node in successors:
+                if child_node in self.closed_nodes:
+                    continue
+
+                if child_node not in self.open_nodes:
+                    self.open_nodes.append(child_node)
+                else:
+                    # retrieve the best current path
+                    better_node = self.open_nodes.pop(self.open_nodes.index(child_node))
+
+                    if child_node.g_cost < better_node.g_cost:
+                        self.open_nodes.append(child_node)
+                    else:
+                        self.open_nodes.append(better_node)
+
+        if not (self.reached):
+            print("No path found")
+
+    def get_successors(self, parent: Node) -> List[Node]:
+        """
+        Returns a list of successors (both in the grid and free spaces)
+        """
+        successors = []
+        for action in delta:
+            pos_x = parent.pos_x + action[1]
+            pos_y = parent.pos_y + action[0]
+            if (
+                pos_x < 0
+                or pos_x > len(grid[0]) - 1
+                or pos_y < 0
+                or pos_y > len(grid) - 1
+            ):
```

Comment:
```suggestion
            if not (0 < pos_x < len(grid[0]) - 1 and 0 < pos_y < len(grid) - 1):
```
Impressive little optimization...  Compare the before / after times!

Reply:
:+1: 

**Before** : 
```
AStar execution time =  0.22050976753234863  seconds
BidirectionalAStar execution time =  0.0417177677154541  seconds
```

**After** :
```
AStar execution time =  2.1457672119140625e-05  seconds
BidirectionalAStar execution time =  1.52587890625e-05  seconds
```